nexus/src/db/explain.rs

```diff
@@ -0,0 +1,251 @@
+// output, which would be possible by calling "get_result" instead of
+// "get_results". We'd like to be able to constrain callers such that they get
+// all of the output or none of it.
+//
+// See the [`Explainable`] trait for why this exists.
+struct ExplainStatement<Q> {
+    query: Q,
+}
+
+impl<Q> QueryId for ExplainStatement<Q>
+where
+    Q: QueryId + 'static,
+{
+    type QueryId = ExplainStatement<Q>;
+    const HAS_STATIC_QUERY_ID: bool = Q::HAS_STATIC_QUERY_ID;
+}
+
+impl<Q> Query for ExplainStatement<Q> {
+    type SqlType = diesel::sql_types::Text;
+}
+
+impl<Q> RunQueryDsl<DbConnection> for ExplainStatement<Q> {}
+
+impl<Q> QueryFragment<Pg> for ExplainStatement<Q>
+where
+    Q: QueryFragment<Pg>,
+{
+    fn walk_ast(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        out.push_sql("EXPLAIN (");
+        self.query.walk_ast(out.reborrow())?;
+        out.push_sql(")");
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use crate::db;
+    use async_bb8_diesel::{AsyncConnection, AsyncSimpleConnection};
+    use diesel::SelectableHelper;
+    use nexus_test_utils::db::test_setup_database;
+    use omicron_test_utils::dev;
+    use uuid::Uuid;
+
+    mod schema {
+        use diesel::prelude::*;
+
+        table! {
+            test_users {
+                id -> Uuid,
+                age -> Int8,
+                height -> Int8,
+            }
+        }
+    }
+
+    use schema::test_users;
+
+    #[derive(Clone, Debug, Queryable, Insertable, PartialEq, Selectable)]
+    #[table_name = "test_users"]
+    struct User {
+        id: Uuid,
+        age: i64,
+        height: i64,
+    }
+
+    async fn create_schema(pool: &db::Pool) {
+        pool.pool()
+            .get()
+            .await
+            .unwrap()
+            .batch_execute_async(
+                "CREATE TABLE test_users (
+                id UUID PRIMARY KEY,
+                age INT NOT NULL,
+                height INT NOT NULL
+            )",
+            )
+            .await
+            .unwrap();
+    }
+
+    // Tests the ".explain()" method in a synchronous context.
+    //
+    // This is often done when calling from transactions, which we demonstrate.
+    #[tokio::test]
+    async fn test_explain() {
+        let logctx = dev::test_setup_log("test_explain");
+        let db = test_setup_database(&logctx.log).await;
+        let cfg = db::Config { url: db.pg_config().clone() };
+        let pool = db::Pool::new(&cfg);
+
+        create_schema(&pool).await;
+
+        use schema::test_users::dsl;
+        pool.pool()
+            .transaction(move |conn| -> Result<(), db::error::TransactionError<()>> {
+                let explanation = dsl::test_users
+                    .filter(dsl::id.eq(Uuid::nil()))
+                    .select(User::as_select())
+                    .explain(conn)
+                    .unwrap();
+                assert_eq!(r#"distribution: local
```

Comment:
I imagine this is likely to change with Cockroach versions.  Worth putting this into a file we can use expectorate with?

Reply:
Updated